systemverilog-plugin/UhdmAst.cc

```diff
@@ -404,58 +405,88 @@ static void add_force_convert_attribute(AST::AstNode *wire_node, int val = 1)
     wire_node->attributes[UhdmAst::force_convert()] = AST::AstNode::mkconst_int(val, true);
 }
 
-static void check_memories(AST::AstNode *module_node)
+static std::map<std::string, AST::AstNode *> check_memories(AST::AstNode *node, std::list<std::string> scopes,
+                                                            std::map<std::string, AST::AstNode *> memories)
 {
-    std::map<std::string, AST::AstNode *> memories;
-    visitEachDescendant(module_node, [&](AST::AstNode *node) {
-        if (node->str == "\\$readmemh") {
-            if (node->children.size() != 2 || node->children[1]->str.empty() || node->children[1]->type != AST::AST_IDENTIFIER) {
-                log_error("%s:%d: Wrong usage of '\\$readmemh'\n", node->filename.c_str(), node->location.first_line);
-            }
-            if (memories[node->children[1]->str])
-                add_force_convert_attribute(memories[node->children[1]->str], 0);
-        }
-        if (node->type == AST::AST_WIRE) {
-            const std::vector<AST::AstNode *> packed_ranges =
-              node->attributes.count(UhdmAst::packed_ranges()) ? node->attributes[UhdmAst::packed_ranges()]->children : std::vector<AST::AstNode *>();
-            const std::vector<AST::AstNode *> unpacked_ranges = node->attributes.count(UhdmAst::unpacked_ranges())
-                                                                  ? node->attributes[UhdmAst::unpacked_ranges()]->children
-                                                                  : std::vector<AST::AstNode *>();
-            if (packed_ranges.size() == 1 && unpacked_ranges.size() == 1) {
-                log_assert(!memories.count(node->str));
-                memories[node->str] = node;
+    std::string full_current_scope = "";
+    std::size_t mem;
+    bool mem_found = false;
+
+    if (node->type == AST::AST_GENBLOCK) {
+        scopes.push_back(node->str);
+    }
+
+    for (auto s : scopes) {
```

Comment:
Maybe lets already store full scope in scopes, so this won't be needed?

Reply:
Right, good point.

Reply:
I reworked the changes, and now we work on a scope string only instead of a list of scopes.